daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TailCallTest.scala

```diff
@@ -0,0 +1,150 @@
+    expectThrows {
+      runExpr(exp, envBound = small, kontBound = unbounded)
+    }
+    expectThrows {
+      runExpr(exp, envBound = unbounded, kontBound = small)
+    }
+  }
+
+  "A tail-recursive definition executes with a small env-stack, and a small kont-stack" in {
+    val exp = e"F:triangleTR 100"
+    val expected = SValue.SInt64(5050)
+    runExpr(exp, envBound = small, kontBound = small) shouldBe expected
+  }
+
+  "fold-left executes with a small env-stack, and a small kont-stack" in {
+    val exp = e"F:triangle_viaFoldLeft 100"
+    val expected = SValue.SInt64(5050)
+    runExpr(exp, envBound = small, kontBound = small) shouldBe expected
+  }
+
+  "fold-right executes with a small env-stack, and a small kont-stack" in {
+    val exp = e"F:triangle_viaFoldRight 100"
+    val expected = SValue.SInt64(5050)
+    runExpr(exp, envBound = small, kontBound = small) shouldBe expected
+  }
+
+  private def typeAndCompile(pkg: Ast.Package): PureCompiledPackages = {
+    val rawPkgs = Map(defaultParserParameters.defaultPackageId -> pkg)
+    Validation.checkPackage(rawPkgs, defaultParserParameters.defaultPackageId, pkg)
+    data.assertRight(
+      PureCompiledPackages(rawPkgs, Compiler.Config.Default.copy(stacktracing = FullStackTrace)))
+  }
+
+  val pkgs = typeAndCompile(pkg)
+
+  private def expectThrows[A](f: => A): Assertion = {
+    val throws: Boolean = try {
+      f
+      false
+    } catch {
+      case _: Throwable => true
+    }
+    throws shouldBe true
+  }
+
+  // Evaluate an expression with optionally bounded env and kont stacks
+  private def runExpr(e: Expr, envBound: Option[Int], kontBound: Option[Int]): SValue = {
+    // create the machine
+    val machine = Speedy.Machine.fromPureExpr(pkgs, e)
+    // maybe replace the env-stack with a bounded version
+    envBound match {
+      case None => ()
+      case Some(bound) =>
+        machine.env = new BoundedArrayList[SValue](bound)
+    }
+    // maybe replace the kont-stack with a bounded version
+    kontBound match {
+      case None => ()
+      case Some(bound) =>
+        val onlyKont: Speedy.Kont =
+          if (machine.kontStack.size != 1) {
+            crash(s"setBoundedKontStack, unexpected size of kont-stack: ${machine.kontStack.size}")
+          } else {
+            machine.kontStack.get(0)
+          }
+        machine.kontStack = new BoundedArrayList[Speedy.Kont](bound)
+        machine.kontStack.add(onlyKont)
+    }
+    // run the machine
+    machine.run() match {
+      case SResultFinalValue(v) => v
+      case res => throw new RuntimeException(s"runExpr, unexpected result $res")
+    }
+  }
+
+  class BoundedArrayList[T](bound: Int) extends util.ArrayList[T](bound) {
+    override def add(x: T): Boolean = {
+      if (size >= bound) {
+        crash(s"BoundedArrayList, exceeded bound of $bound")
+      }
+      super.add(x)
+    }
+  }
+
+  def crash[A](reason: String): A = throw new RuntimeException(reason)
```

Comment:
Don't you do something like this in a language with subtyping?
```suggestion
  def crash(reason: String): Nothing = throw new RuntimeException(reason)
```

Reply:
I think either way is acceptable. The parametric form is used in speedy code `SError.scala`